kernel/queue.c

```diff
@@ -207,19 +207,23 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
 
 	event.state = K_POLL_STATE_NOT_READY;
 
-	err = k_poll(&event, 1, timeout);
-	if (err) {
-		return NULL;
-	}
+	do {
+		err = k_poll(&event, 1, timeout);
+		if (err) {
+			return NULL;
+		}
 
-	__ASSERT_NO_MSG(event.state == K_POLL_STATE_FIFO_DATA_AVAILABLE);
+		__ASSERT_NO_MSG(event.state ==
+				K_POLL_STATE_FIFO_DATA_AVAILABLE);
+
+		/* sys_slist_* aren't threadsafe, so must be always protected by
+		 * irq_lock.
+		 */
+		key = irq_lock();
+		val = sys_slist_get(&queue->data_q);
```

Comment:
if canceled, val will be NULL, does this satisfy the cancelled usecase?

Reply:
If cancelled won't k_poll() return -EINT?

Reply:
k_poll doesn't return -EINT, not sure if there is anything that returns EINT actually.

Reply:
Indeed, not sure how this would work with k_fifo_cancel_wait().

Reply:
So what do we expect to happen when k_fifo_cancel_wait() is called? we expect k_poll() to return err != 0?

Reply:
Well, we expect k_fifo_get() to return with NULL. How that is achieved is an implementation detail, but that case isn't handled by this patch.

Reply:
It is clearly not handling it, weird I thought we had a test for it, apparently not. It is probably because only the fifo tests are testing this which does not select CONFIG_POLL.